perspective.el

```diff
@@ -790,42 +810,45 @@ See also `persp-add-buffer'."
          (old-buffer (window-buffer window)))
     ad-do-it
 
-    (let ((buffer (window-buffer window)))
-      (with-selected-frame frame
-        (unless (memq buffer (persp-buffers (persp-curr)))
-          ;; If a buffer from outside this perspective was selected, it's because
-          ;; this perspective is out of buffers. For lack of any better option, we
-          ;; recreate the scratch buffer.
-          ;;
-          ;; If we were just in a scratch buffer, change the name slightly.
-          ;; Otherwise our new buffer will get deleted too.
-          (let ((name (concat "*scratch* (" (persp-name (persp-curr)) ")")))
-            (when (and bury-or-kill (equal name (buffer-name old-buffer)))
-              (setq name (concat "*scratch*  (" (persp-name (persp-curr)) ")")))
-            (with-selected-window window
-              (switch-to-buffer name)
-              (funcall initial-major-mode))))))))
+    (let ((buffer (window-buffer window))
+          (persp (persp-curr frame)))
+      (unless (memq buffer (persp-buffers persp))
+        ;; If a buffer from outside this perspective was selected, it's because
+        ;; this perspective is out of buffers. For lack of any better option, we
+        ;; recreate the scratch buffer.
+        ;;
+        ;; If we were just in a scratch buffer, change the name slightly.
+        ;; Otherwise our new buffer will get deleted too.
+        (let ((name (concat "*scratch* (" (persp-name persp) ")")))
+          (when (and bury-or-kill (equal name (buffer-name old-buffer)))
+            (setq name (concat "*scratch*  (" (persp-name persp) ")")))
+          (with-selected-window window
+            (switch-to-buffer name)
+            (funcall initial-major-mode)))))))
 
 (defadvice recursive-edit (around persp-preserve-for-recursive-edit)
   "Preserve the current perspective when entering a recursive edit."
   (persp-protect
     (persp-save)
-    (persp-let-frame-parameters ((persp--recursive (persp-curr)))
```

Comment:
We don't need this `persp-let-fram-parameters` since its usage is simple enough that we can just expand out what its doing.